File: raspberry/SmartHome/app/src/main/java/smarthome/raspberry/server/HandlerType.java
package smarthome.raspberry.server;

import android.net.Uri;
import android.util.Log;

import fi.iki.elonen.NanoHTTPD;
import smarthome.raspberry.server.httphandlers.AlertPost;
import smarthome.raspberry.server.httphandlers.ControllerGet;
import smarthome.raspberry.server.httphandlers.ControllerPost;
import smarthome.raspberry.server.httphandlers.ErrorHandler;
import smarthome.raspberry.server.httphandlers.InfoGet;
import smarthome.raspberry.server.httphandlers.InitPost;
import smarthome.raspberry.server.httphandlers.RequestHandler;
import smarthome.raspberry.server.httphandlers.ResetPost;

import static fi.iki.elonen.NanoHTTPD.Method.GET;
import static fi.iki.elonen.NanoHTTPD.Method.POST;

public enum HandlerType {
    READ_CONTROLLER(GET, "/controller", new ControllerGet()),
    CHANGE_CONTROLLER(POST, "/controller", new ControllerPost()),
    INFO(GET, "/info", new InfoGet()),
    INIT(POST, "/init", new InitPost()),
    RESET(POST, "/reset", new ResetPost()),
    ALERT(POST, "/alert", new AlertPost());


    public static final String TAG = HandlerType.class.getSimpleName();
    private static final ErrorHandler errorHandler = new ErrorHandler();
    private RequestHandler handler;
    private NanoHTTPD.Method method;
    private String requestPath;

    HandlerType(NanoHTTPD.Method method, String requestPath, RequestHandler handler) {
        this.method = method;
        this.requestPath = requestPath;
        this.handler = handler;
    }

    public static NanoHTTPD.Response handle(NanoHTTPD.IHTTPSession session) {
        NanoHTTPD.Method method = session.getMethod();
        String uri = session.getUri();
        Log.d(TAG, "handle: " + method + ":" + uri + " " + session.getQueryParameterString());

        HandlerType handlerType = findSuitableHandler(method, uri);
        return handlerType.handler.serve(session);
    }

    public static NanoHTTPD.Response errorHandle(String message) {
        return errorHandler.getError(message);
    }

    private static HandlerType findSuitableHandler(NanoHTTPD.Method method, String uri) {
        for (HandlerType handlerType : values()) {
            if (handlerType.method == method && uri.startsWith(handlerType.requestPath)) {
                return handlerType;
            }
        }

        throw new IllegalArgumentException("No handler found for " + method + ", " + uri);
    }
}

File: raspberry/SmartHome/app/src/main/java/smarthome/raspberry/server/HandlerType.kt
package smarthome.raspberry.server

import android.util.Log

import fi.iki.elonen.NanoHTTPD
import smarthome.raspberry.server.httphandlers.AlertPost
import smarthome.raspberry.server.httphandlers.ControllerGet
import smarthome.raspberry.server.httphandlers.ControllerPost
import smarthome.raspberry.server.httphandlers.ErrorHandler
import smarthome.raspberry.server.httphandlers.InfoGet
import smarthome.raspberry.server.httphandlers.InitPost
import smarthome.raspberry.server.httphandlers.RequestHandler
import smarthome.raspberry.server.httphandlers.ResetPost

import fi.iki.elonen.NanoHTTPD.Method.GET
import fi.iki.elonen.NanoHTTPD.Method.POST

enum class HandlerType private constructor(private val method: NanoHTTPD.Method, private val requestPath: String, private val handler: RequestHandler) {
    READ_CONTROLLER(GET, "/controller", ControllerGet()),
    CHANGE_CONTROLLER(POST, "/controller", ControllerPost()),
    INFO(GET, "/info", InfoGet()),
    INIT(POST, "/init", InitPost()),
    RESET(POST, "/reset", ResetPost()),
    ALERT(POST, "/alert", AlertPost());


    companion object {
        val TAG = HandlerType::class.java.simpleName
        private val errorHandler = ErrorHandler()

        suspend fun handle(session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response {
            val method = session.method
            val uri = session.uri
            Log.d(TAG, "handle: " + method + ":" + uri + " " + session.queryParameterString)

            val handlerType = findSuitableHandler(method, uri)
            return handlerType.handler.serve(session)
        }

        fun errorHandle(message: String): NanoHTTPD.Response {
            return errorHandler.getError(message)
        }

        private fun findSuitableHandler(method: NanoHTTPD.Method, uri: String): HandlerType {
            for (handlerType in values()) {
                if (handlerType.method == method && uri.startsWith(handlerType.requestPath)) {
                    return handlerType
                }
            }

            throw IllegalArgumentException("No handler found for $method, $uri")
        }
    }
}
